core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala

```diff
@@ -42,18 +47,169 @@ final class Cleanup(systemProvider: ClassicActorSystemProvider, settings: Cleanu
+
+    }
+  }
+
+  /**
+   * Keep all snapshots that occurred after `keepAfter`.
+   * If fewer than `snapshotsToKeep` occurred after `keepAfter` at least that many
+   * are kept. Setting this to 1 ensures that at least snapshot is kept even if it
+   * is older than the `keepAfter`
+   *
+   * If only N number of snapshot should be kept prefer overload without timestamp
+   * as it is more efficient.
+   *
+   * The returned snapshot metadata can be used to issue deletes for events older than the oldest
+   * snapshot.
+   *
+   * @return the snapshot meta of the oldest remaining snapshot. None if there are no snapshots
+   *
+   */
+  def deleteBeforeSnapshot(
+      persistenceId: String,
+      snapshotsToKeep: Int,
+      keepAfter: Long): Future[Option[SnapshotMetadata]] = {
+    require(snapshotsToKeep >= 1, "must keep at least one snapshot")
+    require(keepAfter >= 0, "keepAfter must be greater than 0")
+    selectAllSnapshotMetaPs
+      .flatMap { ps =>
+        val allRows: Source[Row, NotUsed] = session.select(ps.bind(persistenceId))
+        @volatile var count = 0
+        allRows
+          .takeWhile { row =>
+            if (row.getLong("timestamp") > keepAfter) {
+              count += 1
+              true
+            } else if (count < snapshotsToKeep) {
+              count += 1
+              true
+            } else {
+              false
+            }
+          }
+          .runWith(Sink.seq)
+      }
+      .flatMap(rows => issueSnapshotDelete(persistenceId, snapshotsToKeep, rows))
+  }
+
+  /**
+   * Keep N snapshots and delete all older snapshots along.
+   *
+   * This operation is much cheaper than including the timestamp because it can use the primary key and limit.
+   *
+   * @return the snapshot meta of the oldest remaining snapshot. None if there are no snapshots. This can be used to delete events from before the snapshot.
+   */
+  def deleteBeforeSnapshot(persistenceId: String, maxSnapshotsToKeep: Int): Future[Option[SnapshotMetadata]] = {
+    require(maxSnapshotsToKeep >= 1, "Must keep at least one snapshot")
+    val snapshots: Future[immutable.Seq[Row]] = selectLatestSnapshotsPs.flatMap { ps =>
+      session.select(ps.bind(persistenceId, maxSnapshotsToKeep: JInt)).runWith(Sink.seq)
+    }
+    snapshots.flatMap(rows => issueSnapshotDelete(persistenceId, maxSnapshotsToKeep, rows))
+  }
+
+  /**
+   * Delete all events before a sequenceNr for the given persistence id.
+   * @param persistenceId the persistence id to delete for
+   * @param toSequenceNr sequence nr (inclusive) to delete up to
+   */
+  def deleteTo(persistenceId: String, toSequenceNr: Long): Future[Done] = {
+    (journal ? CassandraJournal.DeleteEventsTo(persistenceId, toSequenceNr)).mapTo[Done]
+  }
+
+  /**
+   * Deletes all but the last N snapshots and deletes all events before this snapshot
+   * Does not delete from the tag_views table
```

Comment:
Perhaps additional caveat about consequences for this? Events by tag would fail if you run it and see a tag for a message that is not there?

Reply:
I don't think it'll fail, it doesn't reconcile that way around. Only messages to tag_views table. I will add warnings tho